substrate/frame/support/src/traits/misc.rs

```diff
@@ -38,6 +41,22 @@ pub const DEFENSIVE_OP_PUBLIC_ERROR: &str = "a defensive failure has been trigge
 #[doc(hidden)]
 pub const DEFENSIVE_OP_INTERNAL_ERROR: &str = "Defensive failure has been triggered!";
 
+pub trait HasDecodeWithMemTrackingCompact: HasCompact<Type = Self::BoundedType> {
+	type BoundedType: for<'a> EncodeAsRef<'a, Self>
+		+ Decode
+		+ From<Self>
+		+ Into<Self>
+		+ DecodeWithMemTracking;
+}
+
+impl<T> HasDecodeWithMemTrackingCompact for T
+where
+	T: HasCompact,
+	<T as HasCompact>::Type: DecodeWithMemTracking,
+{
+	type BoundedType = <T as HasCompact>::Type;
+}
+
```

Comment:
would this work? it would avoid an unnecessary associated type.
```suggestion
pub trait HasDecodeWithMemTrackingCompact: HasCompact<Type: DecodeWithMemTracking> {}

impl<T> HasDecodeWithMemTrackingCompact for T
where
	T: HasCompact<Type: DecodeWithMemTracking>,
{}

```

Reply:
Nice ! It works. Done.

I just realized one more thing about `HasDecodeWithMemTrackingCompact`. We need to make this trait public and the customers might start using it. In the future, if we remove it, it will be a breaking change again. This would be one more reason why I would prefer to not add it and add `where` bounds when needed. Just wanted to mention this, but no strong preferences.

Reply:
I see, I still prefer to have reasonable constraint on the type rather than `where` bounds.
The idea is that we are already very generic, we don't need to make some FRAME tool usable without `Block` implementing `DecodeWithMemTracking` if FRAME requires it for its standard runtime definition.

But if other people have strong opinion in the other direction I will also accept it.

Reply:
Also instead of using the new automatically implemented trait `HasDecodeWithMemTrackingCompact` we coud bound `HasCompact<Type: DecodeWithMemTracking> ` directly when we need it. But using a trait to easily bound this is ok to me.

Reply:
> Also instead of using the new automatically implemented trait `HasDecodeWithMemTrackingCompact` we coud bound `HasCompact<Type: DecodeWithMemTracking> ` directly when we need it.

Yes, I think this is better. Done.



> I see, I still prefer to have reasonable constraint on the type rather than `where` bounds.
> The idea is that we are already very generic, we don't need to make some FRAME tool usable without `Block` implementing `DecodeWithMemTracking` if FRAME requires it for its standard runtime definition.

Makes sense.